File: skellig-test-step-processing-http/src/main/java/org/skellig/teststep/processor/http/model/HttpTestStep.java
package org.skellig.teststep.processor.http.model;

import org.skellig.teststep.processing.model.TestStep;
import org.skellig.teststep.processing.model.TestStepExecutionType;
import org.skellig.teststep.processing.model.ValidationDetails;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;

public class HttpTestStep extends TestStep {

    private String url;
    private Collection<String> services;
    private String method;
    private String username;
    private String password;
    private Map<String, String> headers;
    private Map<String, String> query;
    private Map<String, String> form;

    protected HttpTestStep(String id, String name, TestStepExecutionType execution, int timeout, int delay,
                           Map<String, Object> variables, Object testData, ValidationDetails validationDetails,
                           String url, Collection<String> services, String method, String username, String password,
                           Map<String, String> headers, Map<String, String> query, Map<String, String> form) {
        super(id, name, execution, timeout, delay, variables, testData, validationDetails);
        this.url = url;
        this.services = services;
        this.method = method;
        this.username = username;
        this.password = password;
        this.headers = headers;
        this.query = query;
        this.form = form;
    }

    public String getUrl() {
        return url;
    }

    public Collection<String> getServices() {
        return services;
    }

    public String getMethod() {
        return method;
    }

    public String getUsername() {
        return username;
    }

    public String getPassword() {
        return password;
    }

    public Map<String, String> getHeaders() {
        return headers;
    }

    public Map<String, String> getQuery() {
        return query;
    }

    public Map<String, String> getForm() {
        return form;
    }

    public static class Builder extends TestStep.Builder {

        private Collection<String> services = Collections.emptyList();
        private String url;
        private String method;
        private String username;
        private String password;
        private Map<String, String> headers;
        private Map<String, String> query;
        private Map<String, String> form;


        public Builder withService(Collection<String> services) {
            this.services = services;
            return this;
        }

        public Builder withUrl(String url) {
            this.url = url;
            return this;
        }

        public Builder withMethod(String method) {
            this.method = method;
            return this;
        }

        public Builder withUsername(String username) {
            this.username = username;
            return this;
        }

        public Builder withPassword(String password) {
            this.password = password;
            return this;
        }

        public Builder withHeaders(Map<String, String> headers) {
            this.headers = headers;
            return this;
        }

        public Builder withQuery(Map<String, String> query) {
            this.query = query;
            return this;
        }

        public Builder withForm(Map<String, String> form) {
            this.form = form;
            return this;
        }

        @Override
        public HttpTestStep build() {
            return new HttpTestStep(id, name, execution, timeout, delay, variables, testData, validationDetails,
                    url, services, method, username, password, headers, query, form);
        }
    }
}

File: skellig-test-step-processing-http/src/main/kotlin/org/skellig/teststep/processor/http/model/HttpTestStep.kt
package org.skellig.teststep.processor.http.model

import org.skellig.teststep.processing.model.DefaultTestStep
import org.skellig.teststep.processing.model.ScenarioStateUpdater
import org.skellig.teststep.processing.model.TestStepExecutionType
import org.skellig.teststep.processing.model.ValidationNode
import org.skellig.teststep.processing.util.PropertyFormatUtils
import org.skellig.teststep.processing.util.PropertyFormatUtils.Companion.createIndent

/**
 * Represents an HTTP test step used to execute HTTP requests from [HttpTestStepProcessor][org.skellig.teststep.processor.http.HttpTestStepProcessor].
 *
 * @param id The unique identifier of the test step
 * @param name The name of the test step
 * @param execution The type of test step execution
 * @param timeout The timeout duration for the test step
 * @param delay The delay duration before executing the test step
 * @param attempts The number of attempts to execute the test step
 * @param values The map of variable names and their corresponding values
 * @param testData The test data for the test step
 * @param validationDetails The validation details for the test step
 * @param scenarioStateUpdaters The list of scenario state updaters for the test step
 * @param url The URL for the HTTP request
 * @param services The collection of services for the test step
 * @param method The HTTP method to be used for the request
 * @param username The username for authentication
 * @param password The password for authentication
 * @param headers The map of request headers
 * @param query The map of query parameters
 * @param form The map of form data
 */
class HttpTestStep(
    id: String?,
    name: String,
    execution: TestStepExecutionType?,
    timeout: Int,
    delay: Int,
    attempts: Int,
    values: Map<String, Any?>?,
    testData: Any?,
    validationDetails: ValidationNode?,
    scenarioStateUpdaters: List<ScenarioStateUpdater>?,
    val url: String?,
    val services: Collection<String>?,
    val method: String?,
    val username: String?,
    val password: String?,
    val headers: Map<String, String?>?,
    val query: Map<String, String?>?,
    val form: Map<String, String?>?
) : DefaultTestStep(id, name, execution, timeout, delay, attempts, values, testData, validationDetails, scenarioStateUpdaters) {

    override fun toString(): String {
        return super.toString() +
                "services = $services\n" +
                "url = $url\n" +
                "method = $method\n" +
                (username?.let { "username = $username\n" } ?: "") +
                (password?.let { "password = $password\n" } ?: "") +
                headers?.let { it.entries.joinToString("\n", "headers {\n", "\n}\n") { n -> "${createIndent(1)}$n" } } +
                query?.let { it.entries.joinToString("\n", "query {\n", "\n}\n") { n -> "${createIndent(1)}$n" } } +
                form?.let { it.entries.joinToString("\n", "form {\n", "\n}\n") { n -> "${createIndent(1)}$n" } }
    }

    class Builder : DefaultTestStep.Builder<HttpTestStep>() {

        private var services: Collection<String>? = emptyList()
        private var url: String? = null
        private var method: String? = null
        private var username: String? = null
        private var password: String? = null
        private var headers: Map<String, String?>? = null
        private var query: Map<String, String?>? = null
        private var form: Map<String, String?>? = null

        fun withService(services: Collection<String>?) = apply {
            this.services = services
        }

        fun withUrl(url: String?) = apply {
            this.url = url
        }

        fun withMethod(method: String?) = apply {
            this.method = method
        }

        fun withUsername(username: String?) = apply {
            this.username = username
        }

        fun withPassword(password: String?) = apply {
            this.password = password
        }

        fun withHeaders(headers: Map<String, String?>?) = apply {
            this.headers = headers
        }

        fun withQuery(query: Map<String, String?>?) = apply {
            this.query = query
        }

        fun withForm(form: Map<String, String?>?) = apply {
            this.form = form
        }

        override fun build(): HttpTestStep {
            return HttpTestStep(
                id, name!!, execution, timeout, delay, attempts, values, testData, validationDetails,
                scenarioStateUpdaters, url, services, method, username, password, headers, query, form
            )
        }
    }
}